go-controller/pkg/retry/obj_retry.go

```diff
@@ -400,10 +402,17 @@ var (
 // free its resources. (for now, this applies to completed pods)
 // processObjectInTerminalState doesn't unlock key
 func (r *RetryFramework) processObjectInTerminalState(obj interface{}, lockedKey string, event resourceEvent) {
+	_, loaded := r.terminatedObjects.LoadOrStore(lockedKey, true)
+	if loaded {
+		// object was already terminated
+		klog.Infof("Detected object %s of type %s in terminal state (e.g. completed) will be " +
+			"ignored as it has already been processed")
+		return
+	}
```

Comment:
If delete fails, it will be added to the retry queue, so further event should be ignored, and the first one that attempted to delete will retry (hopefully I got your quest right)

Reply:
OK I was confused because I thought a delete event would replace an update event, but this change actually prevents that from happening. So it would be the update event which finally handles the delete in this case. That means that we need to make sure they are actually equivalent. So basically `processObjectInTerminalState` and the delete handler should share the code. 

There is one difference now, `InitRetryObjWithDelete` is called with a different value for the last boolean in this both cases. I am nto sure how meaningful that is.

Reply:
its probably ok...that boolean seems to have to be false in cases where we are deleting to then process an add or an update within the same event. And it should be true when we just want to delete and nothing else.
So it would probably have been more correct for it to be true in the delete handler, instead of false, but it would not matter anyway in a delete event because that's just the only thing we want to do.

Reply:
@trozet another difference are the events recorded. In `processObjectInTerminalState` we record a error event but no success event. In the delete handler is the other way around, we record the success event but not the error event. Which made sense to be that way but now it means that we actually might lose the delete success events.

Reply:
we should consolidate and fix this delete code to a single function like you mentioned over slack. Somebody can take that up in another PR